src/sentry/static/sentry/app/views/groupDetails/shared/groupEventDetails.jsx

```diff
@@ -37,14 +39,39 @@ class GroupEventDetails extends React.Component {
     this.fetchData();
   }
 
-  componentDidUpdate(prevProps) {
-    if (prevProps.params.eventId !== this.props.params.eventId) {
+  componentDidUpdate(prevProps, prevState) {
+    const {environments, params} = this.props;
+
+    const eventHasChanged = prevProps.params.eventId !== params.eventId;
+    const environmentsHaveChanged = !isEqual(prevProps.environments, environments);
+
+    // If environments are being actively changed and will no longer contain the
+    // current event's environment, redirect to latest
+    if (
+      environmentsHaveChanged &&
+      prevState.event &&
+      params.eventId &&
+      !['latest', 'oldest'].includes(params.eventId)
+    ) {
+      const shouldRedirect =
+        environments.length > 0 &&
+        !environments.find(env => env.name === getEventEnvironment(prevState.event));
+
+      if (shouldRedirect) {
+        browserHistory.replace(
+          `/organizations/${params.orgId}/issues/${params.groupId}/`
```

Comment:
Do you need to retain an of the query string parameters?

Reply:
Yeah that'd be better - updated